### ISSUE TITLE
Semantic Tokens: Distinguish (possibly-partially-applied) infix operators from normal functions/variables

### ISSUE THREAD

**Issue by konn** [haskell/haskell-language-server#3969]
## Is your enhancement request related to a problem? Please describe.

First of all, thank you @soulomoon for implementating fantastic support for semantic tokens! I've long been waiting for this feature :tada:

After the release of 2.6.0.0, I've just tested it and looks really good. I think it will become even better if one can tell (possibly partially-applied) infix operators from ordinary variable/functions semantically.

For example, the image below is what I apply One Monokai + Semantic Tokens to the code fragments containing several `($)`s:

![code-semantic](https://github.com/haskell/haskell-language-server/assets/21860/a3062658-d177-4e67-88c6-d9894c8a120f)

In above, functions being applied and `($)`s are all coloured homogeneously. Without semantic colouring, OTOH, we got:

![code-syntactic](https://github.com/haskell/haskell-language-server/assets/21860/e81b44fb-173c-4259-9333-6226596826ee)

I think the latter case gives more legible highlighting - binary operator shines between functions.

## Describe the solution you'd like

Add something like `TBinaryOperator` constructor to `HsSemanticTokenType`, map its default to `operator` token, and also allowing users to map them to what they like (as we already did for ` haskell.plugin.semanticTokens.config.dataConstructorToken` etc).

## Describe alternatives you've considered

Perhaps, as we have already done to plain symbols, there can be `TBinaryFunction` and `TBinaryVariable`, as the function declaration can take its argument in infix-form.

**Comment by soulomoon** [comment]
Thank you @konn for ideas! I've added it to the tracklist!

**Comment by michaelpj** [comment]
We even have a good LSP token type for this: `operator`.

**Comment by michaelpj** [comment]
Also, what about ` `isSubsetOf` ` vs `+`? Is the former also an operator or should it continue to be a function?

**Comment by konn** [comment]
> We even have a good LSP token type for this: `operator`.

Indeed, as I already suggested in my solution section 😃 

> Also, what about `` `isSubsetOf` `` vs `+`? Is the former also an operator or should it continue to be a function?

Good catch. Both seem making sense, but I prefer treating it as a function.

**Comment by michaelpj** [comment]
Yep. I think you already have support for returning multiple types and picking one.

I note that there is a client capability for supporting overlapping tokens. I guess in principle that means we could return two tokens for the same span? I don't know what that would do.

**Comment by michaelpj** [comment]
I agree

**Comment by konn** [comment]
> > I note that there is a client capability for supporting overlapping tokens
> 
> I suspect it is scarcely supported by clients. e.g. Vscode decided not to support this. [microsoft/vscode#147774](https://github.com/microsoft/vscode/issues/147774)
> 
> I am inclined to take precedency of `class method` over `infix operator` when it comes into conflict. What are you guys opinions on this?

I prefer it to be infix operator, but it is just a matter of taste. We might be able to make such precedence preference configurable, but that seems rather complex and non-ergonomic. After all, humans can tell the infix operators from plain class method as it consists of (non-letter) symbols. So, yes, treating it as class method seems making sense to me.

**Comment by soulomoon** [comment]
One argument is that:

I think the in most case, operator is needed since we want to differ them from

**Comment by michaelpj** [comment]
We have several clashes already, in fact:
- `function` vs not
- `operator` vs not
- `method` vs not

Something can be all of a function, an operator, and a method!

On the configurability front, we could consider making it so that we don't take certain things into account. So e.g. you could turn off highlighting methods entirely, with the intention that you now get the next token assignation that makes sense. 

Alternatively we could expose priority options for the different token types (as integers), and let people change them.

**Comment by soulomoon** [comment]
#4030 should fix this now, give a try if it is working well? @konn

**Comment by konn** [comment]
@soulomoon Thanks! And, sorry for my late reply (I had been a bit busy for the past two weeks). I will give it a try.

**Comment by konn** [comment]
Looks good! Here is the example (I switched the colour theme from One Monokai to Material Theme Ocean, as the former doesn't work well with semantic tokens for the time being).

_Without_ semantic tokens:

![code-plain](https://github.com/haskell/haskell-language-server/assets/21860/1550d1ad-139f-4e96-97f8-6b9858d7a4a7)

... and _with_ semantic tokens:

![code-semantic](https://github.com/haskell/haskell-language-server/assets/21860/703ad99b-6fef-4cf8-b794-8c0d9b834653)

Now we can distinguish binary operators from functions! Looks really good 😄 

I also tested `` `func` ``-style infixes:

_Without_ semantic tokens:

![code2-plain](https://github.com/haskell/haskell-language-server/assets/21860/e7bac2b0-c4a3-4dd4-a2c9-32202dda4104)

... and _with_ semantic tokens:

![code2-semantic](https://github.com/haskell/haskell-language-server/assets/21860/62e97369-e402-47e8-9313-898ce128e977)

The above example has both symbolic and `` `func` ``-style (backticked) operators. 
Without semantic tokens, default syntax defs treat backticked operators, as a whole, as a kind of infix operators.
With semantic tokens, the `quot` function itself is highlighted as a function, and a backtick as a punctuation.

language | haskell
-- | --
standard token type | Other
foreground | #89DDFF
background | #0F111A
contrast ratio | 12.41

item |  value
-- | --
textmate scopes | `punctuation.backtick.haskell`, `keyword.operator.function.infix.haskell`, `source.haskell`
foreground | punctuation `{ "foreground": "#89DDFF" }`

I think the current treatment is good enough - after all, `quot` itself is a function and what makes it infix-style is `` ` `` and they are treated differently, indicating the details of how the things work.
